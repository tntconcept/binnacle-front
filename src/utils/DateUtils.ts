import {
  addWeeks,
  eachDayOfInterval,
  endOfMonth,
  endOfWeek,
  format,
  isSameDay,
  parse,
  startOfMonth,
  startOfWeek,
  subWeeks
} from 'date-fns'
import { es } from 'date-fns/locale'
import { IHolidays } from 'api/interfaces/IHolidays'

export const formatDateForQuery = (date: Date) => format(date, 'yyyy-MM-dd')

export const firstDayOfFirstWeekOfMonth = (month: Date) => {
  return startOfWeek(startOfMonth(month), { weekStartsOn: 1 })
}

export const lastDayOfLastWeekOfMonth = (month: Date) => {
  return endOfWeek(endOfMonth(month), { weekStartsOn: 1 })
}

export const getDatesIntervalByMonth = (selectedMonth: Date) => {
  return eachDayOfInterval({
    start: firstDayOfFirstWeekOfMonth(selectedMonth),
    end: lastDayOfLastWeekOfMonth(selectedMonth)
  })
}

export const getDaysOfWeek = (start: Date) => {
  return eachDayOfInterval({
    start: startOfWeek(start, { weekStartsOn: 1 }),
    end: endOfWeek(start, { weekStartsOn: 1 })
  })
}
export const getPreviousWeek = (week: Date) => {
  return subWeeks(week, 1)
}
export const getNextWeek = (week: Date) => {
  return addWeeks(week, 1)
}

export const formatDayAndMonth = (date: Date) => {
  const isSpanish = navigator.language === 'es-ES'
  const locale = isSpanish ? es : undefined
  const dateFormat = isSpanish ? "dd 'de' MMMM" : 'dd MMMM'
  return format(date, dateFormat, { locale })
}

const getUTCDate = (dateString = Date.now()) => {
  const date = new Date(dateString)

  return new Date(
    date.getUTCFullYear(),
    date.getUTCMonth(),
    date.getUTCDate(),
    date.getUTCHours(),
    date.getUTCMinutes(),
    date.getUTCSeconds()
  )
}

export const timeToDate = (time: string, backupDate?: Date) => {
  return parse(time, 'HH:mm', backupDate || getUTCDate())
}

export const getWeekdaysName = () => {
  const isSpanish = navigator.language === 'es-ES'
  if (isSpanish) {
    return ['lun', 'mar', 'mié', 'jue', 'vie', 'sáb', 'dom']
  }
  return ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun']
}
export const isHoliday = (publicHolidays: IHolidays['holidays'], date: Date) =>
  publicHolidays.find((holiday) => isSameDay(holiday.date, date))

export const isVacation = (privateHolidays: IHolidays['vacations'], date: Date) =>
  privateHolidays.find((vacation) => vacation.days.some((day) => isSameDay(day, date)))
