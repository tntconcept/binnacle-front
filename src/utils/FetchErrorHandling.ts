import i18n from "../i18n"
import {WretcherError} from "wretch"

// TODO Refactor key que sea un string y acepte unknown y offline, ademÃ¡s de que devuelva la key.
interface ICustomStatusMessages {
  [key: string]: {
    title: string;
    description: string;
  };
}

const statusCodeMap: ICustomStatusMessages = {
  "401": {
    title: i18n.t("api_errors.unauthorized"),
    description: i18n.t("api_errors.unauthorized_description")
  },
  "403": {
    title: i18n.t("api_errors.forbidden"),
    description: i18n.t("api_errors.forbidden_description")
  },
  "404": {
    title: i18n.t("api_errors.not_found"),
    description: i18n.t("api_errors.general_description")
  },
  "408": {
    title: i18n.t("api_errors.timeout"),
    description: i18n.t("api_errors.general_description")
  },
  "500": {
    title: i18n.t("api_errors.server_error"),
    description: i18n.t("api_errors.general_description")
  },
  "503": {
    title: i18n.t("api_errors.server_down"),
    description: i18n.t("api_errors.general_description")
  },
  "offline": {
    title: i18n.t("api_errors.offline"),
    description: i18n.t("api_errors.offline_description")
  },
  "unknown": {
    title: i18n.t("api_errors.unknown"),
    description: i18n.t("api_errors.general_description")
  },
};

const getTimeoutOrUnknownStatusCode = (error: WretcherError) => {
  if (!navigator.onLine) {
    return "offline"
  }

  if (error.name === "AbortError") {
    return 408;
  } else if (!error.response) {
    // Network error -> Server is down
    return 503;
  } else {
    // Unknown error, CORS maybe
    return "unknown";
  }
};

const getMessageByStatusCode = (
  error: WretcherError,
  customStatusCodeMessages: ICustomStatusMessages = {}
) => {
  const statusCode =
    error.response === undefined
      ? getTimeoutOrUnknownStatusCode(error)
      : error.response.status;

  const text = { ...statusCodeMap, ...customStatusCodeMessages };

  return text[statusCode] || text["unknown"];
};

const getErrorMessage = (
  error: WretcherError,
  overrideStatusCodeMessages: ICustomStatusMessages = {}
) => {
  return getMessageByStatusCode(error, overrideStatusCodeMessages)
};

export default getErrorMessage;
