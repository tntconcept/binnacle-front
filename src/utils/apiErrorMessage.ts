import axios, { AxiosError } from "axios";
import i18n from "../i18n";

interface ICustomStatusMessages {
  [key: number]: {
    title: string;
    description: string;
  };
}

const statusCodeMap: ICustomStatusMessages = {
  401: {
    title: i18n.t("api_errors.unauthorized"),
    description: i18n.t("api_errors.unauthorized_description")
  },
  403: {
    title: i18n.t("api_errors.forbidden"),
    description: i18n.t("api_errors.forbidden_description")
  },
  404: {
    title: i18n.t("api_errors.not_found"),
    description: i18n.t("api_errors.general_description")
  },
  408: {
    title: i18n.t("api_errors.timeout"),
    description: i18n.t("api_errors.general_description")
  },
  500: {
    title: i18n.t("api_errors.server_error"),
    description: i18n.t("api_errors.general_description")
  },
  503: {
    title: i18n.t("api_errors.server_down"),
    description: i18n.t("api_errors.general_description")
  },
  600: {
    title: i18n.t("api_errors.unknown"),
    description: i18n.t("api_errors.general_description")
  }
};

const getTimeoutOrUnknownStatusCode = (error: AxiosError) => {
  // Axios timeouts the request because was longer than the seconds established.
  if (error.code === "ECONNABORTED") {
    return 408;
  } else if (!error.response) {
    // Network error -> Server is down
    return 503;
  } else {
    // Unknown error, CORS maybe
    return 600;
  }
};

const getMessageByStatusCode = (
  error: AxiosError,
  customStatusCodeMessages: ICustomStatusMessages = {}
) => {
  const statusCode =
    error.response === undefined
      ? getTimeoutOrUnknownStatusCode(error)
      : error.response.status;

  const text = { ...statusCodeMap, ...customStatusCodeMessages };

  return text[statusCode] || text[600];
};

const getErrorMessage = (
  error: AxiosError,
  overrideStatusCodeMessages: ICustomStatusMessages = {}
) => {
  if (!axios.isCancel(error)) {
    return getMessageByStatusCode(error, overrideStatusCodeMessages);
  } else {
    if (process.env.NODE_ENV !== "production") {
      let messageLog: string;

      if (error.message) {
        messageLog = `, message:: ${error.message}`;
      } else {
        messageLog = "without any message";
      }

      // eslint-disable-next-line
      console.log(`The request was cancelled ${messageLog}`);
    }
  }
};

export default getErrorMessage;
