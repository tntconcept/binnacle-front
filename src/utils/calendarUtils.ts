import {
  eachDayOfInterval,
  endOfMonth,
  endOfWeek,
  startOfMonth,
  startOfWeek
} from "date-fns";

export const getDatesIntervalByMonth = (selectedMonth: Date) => {
  return eachDayOfInterval({
    start: startOfWeek(startOfMonth(selectedMonth), { weekStartsOn: 1 }),
    end: endOfWeek(endOfMonth(selectedMonth), { weekStartsOn: 1 })
  });
};
