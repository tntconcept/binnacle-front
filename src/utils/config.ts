// Don't forget these APIs are experimental and can change.
export const suspenseConfig = {
  timeoutMs: 6000,
  busyDelayMs: 500, // Before we show the inline spinner
  busyMinDurationMs: 100, // If we show it, force it to stick for a bit
};