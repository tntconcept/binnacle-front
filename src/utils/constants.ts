export const SUSPENSE_CONFIG = {
  timeoutMs: 6000,
  busyDelayMs: 500, // Before we show the inline spinner <- Deprecated now is done using css.
  busyMinDurationMs: 100 // If we show it, force it to stick for a bit
}
