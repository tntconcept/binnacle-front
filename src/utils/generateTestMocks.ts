import { IOAuth } from 'api/interfaces/IOAuth'
import { IActivity } from 'api/interfaces/IActivity'
import { IOrganization } from 'api/interfaces/IOrganization'
import { IProjectRole } from 'api/interfaces/IProjectRole'
import { IProject } from 'api/interfaces/IProject'
import { IRecentRole } from 'api/interfaces/IRecentRole'

const generateId = () => {
  return Math.floor(Math.random() * 500)
}

export const buildOAuthResource = (): IOAuth => ({
  access_token: 'test access token',
  token_type: 'bearer',
  refresh_token: 'test refresh token',
  expires_in: 360,
  scope: 'tnt',
  jti: 'jti code'
})

export const buildOrganization = (
  override?: Partial<IOrganization>
): IOrganization => {
  return {
    id: generateId(),
    name: 'Test Organization Name',
    ...override
  }
}

export const buildProject = (override?: Partial<IProject>): IProject => {
  return {
    id: generateId(),
    billable: false,
    name: 'Test Project Name',
    open: true,
    ...override
  }
}

export const buildProjectRole = (override?: Partial<IProjectRole>): IProjectRole => {
  return {
    id: generateId(),
    name: 'Test Project Role Name',
    requireEvidence: false,
    ...override
  }
}

export const buildRecentRole = (override?: Partial<IRecentRole>): IRecentRole => {
  return {
    id: generateId(),
    requireEvidence: false,
    name: 'Test Recent Role Name',
    date: new Date(),
    projectBillable: false,
    projectName: 'Test Recent Role Project Name',
    organizationName: 'Test Organization Name',
    ...override
  }
}

export const buildActivity = (override?: Partial<IActivity>): IActivity => {
  return {
    id: generateId(),
    billable: false,
    description: 'Lorem Ipsum...',
    startDate: new Date(),
    duration: 100,
    imageFile: '',
    hasImage: false,
    organization: buildOrganization(),
    project: buildProject(),
    projectRole: buildProjectRole(),
    userId: 0,
    ...override
  }
}
