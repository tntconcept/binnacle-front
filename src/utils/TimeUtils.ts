import { roundToTwoDecimals } from 'utils/helpers'
import { addMinutes, format, getMinutes, setMinutes } from 'date-fns'
import DateTime from 'services/DateTime'

export const getDuration = (minutes: number, decimalFormat: boolean = false) => {
  if (decimalFormat) {
    return roundToTwoDecimals(Math.abs(minutes) / 60)
  }

  return Math.abs(minutes) > 0 ? DateTime.getHumanizedDuration(minutes) : '0h'
}

export const getTimeInterval = (startTime: Date, amount: number) => {
  const finalTime = addMinutes(startTime, amount)
  return format(startTime, 'HH:mm') + ' - ' + format(finalTime, 'HH:mm')
}
export const roundHourToQuarters = (date: Date): Date => {
  const roundMinutes = Math.round(getMinutes(date) / 15) * 15
  return setMinutes(date, roundMinutes)
}
