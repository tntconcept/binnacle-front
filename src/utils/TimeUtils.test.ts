import { getDuration, getTimeInterval, roundHourToQuarters } from 'utils/TimeUtils'
import {
  firstDayOfFirstWeekOfMonth,
  getDatesIntervalByMonth,
  lastDayOfLastWeekOfMonth,
  getDaysOfWeek,
  getPreviousWeek,
  getNextWeek
} from 'utils/DateUtils'

describe('Time utils', () => {
  const date = new Date('2020-01-31T20:10:00.000Z')

  it('should format activity time range', function() {
    const result = getTimeInterval(date, 2 * 60)
    expect(result).toEqual('21:10 - 23:10')
  })

  it('should round minute to 15', function() {
    const result1 = roundHourToQuarters(new Date('2020-01-31T20:10:00.000Z'))
    expect(result1).toEqual(new Date('2020-01-31T20:15:00.000Z'))

    const result2 = roundHourToQuarters(new Date('2020-01-31T20:05:00.000Z'))
    expect(result2).toEqual(new Date('2020-01-31T20:00:00.000Z'))

    const result3 = roundHourToQuarters(new Date('2020-01-31T20:00:00.000Z'))
    expect(result3).toEqual(new Date('2020-01-31T20:00:00.000Z'))

    const result4 = roundHourToQuarters(new Date('2020-01-31T20:15:00.000Z'))
    expect(result4).toEqual(new Date('2020-01-31T20:15:00.000Z'))
  })

  it('should format activity duration in decimal format', function() {
    const result1 = getDuration(4 * 60, true)
    expect(result1).toEqual(4)

    const result2 = getDuration(4.25 * 60, true)
    expect(result2).toEqual(4.25)

    const result3 = getDuration(0 * 60, true)
    expect(result3).toEqual(0)
  })

  it('should format activity duration in humanized format', function() {
    const result1 = getDuration(4 * 60, false)
    expect(result1).toEqual('4h')

    const result2 = getDuration(4.25 * 60, false)
    expect(result2).toEqual('4h 15m')

    const result3 = getDuration(0 * 60, false)
    expect(result3).toEqual('0h')
  })

  it('should return the first day of the first week of the month', function() {
    const result = firstDayOfFirstWeekOfMonth(date)
    expect(result).toEqual(new Date('2019-12-29T23:00:00.000Z'))
  })

  it('should return last day of the last week of the month', function() {
    const result = lastDayOfLastWeekOfMonth(date)
    expect(result).toEqual(new Date('2020-02-02T22:59:59.999Z'))
  })

  it('should return each day of month', function() {
    const result = getDatesIntervalByMonth(date)
    expect(result).toMatchSnapshot()
  })

  it('should return each day of week', function() {
    const result = getDaysOfWeek(date)
    expect(result).toMatchSnapshot()
  })

  it('should return previous week', function() {
    const result = getPreviousWeek(date)
    expect(result).toEqual(new Date('2020-01-24T20:10:00.000Z'))
  })

  it('should return next week', function() {
    const result = getNextWeek(date)
    expect(result).toEqual(new Date('2020-02-07T20:10:00.000Z'))
  })
})
