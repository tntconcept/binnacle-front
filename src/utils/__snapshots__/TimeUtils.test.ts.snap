// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time utils should return each day of month 1`] = `
Array [
  2019-12-29T23:00:00.000Z,
  2019-12-30T23:00:00.000Z,
  2019-12-31T23:00:00.000Z,
  2020-01-01T23:00:00.000Z,
  2020-01-02T23:00:00.000Z,
  2020-01-03T23:00:00.000Z,
  2020-01-04T23:00:00.000Z,
  2020-01-05T23:00:00.000Z,
  2020-01-06T23:00:00.000Z,
  2020-01-07T23:00:00.000Z,
  2020-01-08T23:00:00.000Z,
  2020-01-09T23:00:00.000Z,
  2020-01-10T23:00:00.000Z,
  2020-01-11T23:00:00.000Z,
  2020-01-12T23:00:00.000Z,
  2020-01-13T23:00:00.000Z,
  2020-01-14T23:00:00.000Z,
  2020-01-15T23:00:00.000Z,
  2020-01-16T23:00:00.000Z,
  2020-01-17T23:00:00.000Z,
  2020-01-18T23:00:00.000Z,
  2020-01-19T23:00:00.000Z,
  2020-01-20T23:00:00.000Z,
  2020-01-21T23:00:00.000Z,
  2020-01-22T23:00:00.000Z,
  2020-01-23T23:00:00.000Z,
  2020-01-24T23:00:00.000Z,
  2020-01-25T23:00:00.000Z,
  2020-01-26T23:00:00.000Z,
  2020-01-27T23:00:00.000Z,
  2020-01-28T23:00:00.000Z,
  2020-01-29T23:00:00.000Z,
  2020-01-30T23:00:00.000Z,
  2020-01-31T23:00:00.000Z,
  2020-02-01T23:00:00.000Z,
]
`;

exports[`Time utils should return each day of week 1`] = `
Array [
  2020-01-26T23:00:00.000Z,
  2020-01-27T23:00:00.000Z,
  2020-01-28T23:00:00.000Z,
  2020-01-29T23:00:00.000Z,
  2020-01-30T23:00:00.000Z,
  2020-01-31T23:00:00.000Z,
  2020-02-01T23:00:00.000Z,
]
`;
